sqlglot/optimizer/qualify_columns.py

```diff
@@ -422,10 +459,74 @@ def _expand_stars(
             tables = [expression.table]
             _add_except_columns(expression.this, tables, except_columns)
             _add_replace_columns(expression.this, tables, replace_columns)
+
+        elif expression.is_star and (
+            isinstance(expression, exp.Dot)
+            or (isinstance(expression, exp.Column) and expression.table in scope.sources)
+        ):
+            col = expression.find(exp.Column) if isinstance(expression, exp.Dot) else expression
+            starting_struct = str(expression.this) if isinstance(expression.this, exp.Dot) else None
+            structs.append((col, starting_struct))
+
+            # Annotate scope ahead of the struct flattening
+            if not scope_annotated and dialect == "bigquery":
+                annotator.annotate_scope(scope)
+                scope_annotated = True
+
         else:
             new_selections.append(expression)
             continue
 
+        while structs:
+            # [BigQuery] Expand/Flatten foo.bar.* where bar is a struct column
```

Comment:
Should we factor this whole thing out in a separate helper? `_expand_stars` is starting to look quite complicated.